c_binding/libsg.c

```diff
@@ -940,3 +1035,385 @@ int _sg_host_no(char *err_msg, int fd, unsigned int *host_no)
+    char strerr_buff[_LSM_ERR_MSG_LEN];
+    uint8_t sense_key = _T10_SPC_SENSE_KEY_NO_SENSE;
+    struct _ata_registers_input_28_bit *input_registers = NULL;
+
+    assert(err_msg != NULL);
+    assert(fd >= 0);
+    assert(ata_cmd != NULL);
+    assert(ata_output_regs != NULL);
+
+    memset(cdb, 0, ATA_PASS_THROUGH_12_LEN);
+    memset(sense_data, 0, _T10_SPC_SENSE_DATA_MAX_LENGTH);
+    memset(sense_err_msg, 0, _LSM_ERR_MSG_LEN);
+
+    //set BYT_BLOK bit to 1 for all commands
+    cdb[2] = 1 << 2;
+
+    //set CK_COND bit
+    if (need_output_reg)
+        cdb[2] = cdb[2] | (1 << 5);
+
+    //set PROTOCOL value as defined in Table 2, and
+    //set T_LENGTH and T_DIR as appropriate.
+    if (direction == ATA_NON_DATA_COMMAND) {
+        cdb[1] = 3 << 1;
+        cdb[2] = cdb[2] | (1 << 3);
+    } else if (direction == ATA_DATA_IN_COMMAND) {
+        cdb[1] = 4 << 1;
+        cdb[2] = cdb[2] | (1 << 3) | 2;
+    } else if (direction == ATA_DATA_OUT_COMMAND) {
+        cdb[1] = 5 << 1;
+        cdb[2] = cdb[2] | 2;
+    }
+
+    input_registers = (struct _ata_registers_input_28_bit *) ata_cmd;
+
+    cdb[0] = ATA_PASS_THROUGH_12;
+    cdb[3] = input_registers->feature;
+    cdb[4] = input_registers->count;
+    cdb[5] = input_registers->lba_low;
+    cdb[6] = input_registers->lba_mid;
+    cdb[7] = input_registers->lba_high;
+    cdb[8] = input_registers->device;
+    cdb[9] = input_registers->command;
+    cmd_len = ATA_PASS_THROUGH_12_LEN;
+
+    ioctl_errno = _sg_io(fd, cdb, cmd_len, data, data_len, sense_data,
+                         _SG_IO_NO_DATA);
+
+    if (ioctl_errno) {
+        rc = _get_ata_output_regs(sense_data, ata_output_regs);
+
+        if (rc) {
+            _check_sense_data(sense_err_msg, sense_data, &sense_key);
+
+            _lsm_err_msg_set(err_msg, "Got error from SGIO ATA PASSTHROUGH: "
+                             "error %d(%s), %s", ioctl_errno,
+                             strerror_r(ioctl_errno, strerr_buff,
+                                        _LSM_ERR_MSG_LEN),
+                             sense_err_msg);
+        }
+    }
+
+    return rc;
+}
+
+int _sg_log_sense(char *err_msg, int fd, uint8_t page_code,
+                  uint8_t sub_page_code, uint8_t *data)
+{
+    int rc = LSM_ERR_OK;
+    uint8_t tmp_data[_T10_SPC_LOG_SENSE_MAX_LEN];
+    uint8_t cdb[_T10_SPC_LOG_SENSE_CMD_LEN];
+    uint8_t sense_data[_T10_SPC_SENSE_DATA_MAX_LENGTH];
+    int ioctl_errno = 0;
+    char strerr_buff[_LSM_ERR_MSG_LEN];
+    uint8_t sense_key = _T10_SPC_SENSE_KEY_NO_SENSE;
+    char sense_err_msg[_LSM_ERR_MSG_LEN];
+    struct _sg_t10_log_para_hdr *log_hdr = NULL;
+    uint16_t log_data_len = 0;
+
+    assert(err_msg != NULL);
+    assert(fd >= 0);
+    assert(data != NULL);
+
+    memset(sense_err_msg, 0, _LSM_ERR_MSG_LEN);
+    memset(cdb, 0, _T10_SPC_LOG_SENSE_CMD_LEN);
+
+    cdb[0] = LOG_SENSE;
+    cdb[2] = (PAGE_CONTROL_CUMULATIVE_VALS << 6) | (page_code & 0x3f);
+    cdb[3] = sub_page_code & UINT8_MAX;
```

Comment:
@tasleson Example usage: https://github.com/libstorage/libstoragemgmt/blob/master/c_binding/libsg.c#L266

I'll leave this one until @cathay4t comments. From a code perspective, you're correct to say that it's not necessary. I'm indifferent.